Update trainer to the new architecture.

Sopel97 opened the pull request:
This is a slightly cleaned up version of https://github.com/Sopel97/nnue-pytorch/tree/experiment_56, with a few changes:

- weight clipping is moved outside of the optimizer, should behave about the same
- updated visualize.py to at least work, though it doesn't visualize fc layers and the plot for ft weights is slightly off
- updated cross_eval_check.py
- updated perf_sigmoid_fitter.py
- changed LR scheduling based on further results

Now will follow results from individual changes. The separate graphs are not comparable between each other.

- weight clipping 
![plot_14_15](https://user-images.githubusercontent.com/8037982/118631761-18145b80-b7d0-11eb-807b-42503544e753.png)
- per psqt features 
![plot_20_21](https://user-images.githubusercontent.com/8037982/118631033-683eee00-b7cf-11eb-98e2-bc82af30fa2e.png)
- 8 subnets and psqt buckets 
![plot_21_30](https://user-images.githubusercontent.com/8037982/118631200-8b699d80-b7cf-11eb-93c6-fabc782d8698.png)
- MSE over WDL-space evals (not very visible on the graph, but the trend was 3 elo better. Loss not comparable)
![plot_48_49](https://user-images.githubusercontent.com/8037982/118631887-367a5700-b7d0-11eb-83ae-cce7608421be.png)
- disable GC (lower loss)
![plot_53_54](https://user-images.githubusercontent.com/8037982/118632065-5d388d80-b7d0-11eb-92fb-0bd5c3967c02.png)
- drop LR by 0.987 every epoch (trend ~2 elo better, easier parametrization) 
![plot_56_57](https://user-images.githubusercontent.com/8037982/118632249-835e2d80-b7d0-11eb-8859-5d531a7db391.png)

The new architecture diagram:
![HalfKAv2-45056-512x2P8x2 -16-32-1 x8](https://user-images.githubusercontent.com/8037982/118657052-66832380-b7eb-11eb-95a9-030936e30522.png)

vondele commented:
would it make sense to make _default_feature_set_name HalfKAv2^ and the default batch size 16k as well --random-fen-skipping N and --smart-fen-skipping ? 
after that update the readme section `Current recommended training invocation`

Sopel97 commented:
I've updated the defaults (I was using random-fen-skipping=3 and it's sufficient). As for the --smart-fen-skipping, I added --no-smart-fen-skipping that disables it, and made it enabled by default. The old option is left for compatibility but doesn't do anything.

vondele commented:
as we're at it, set the default number of threads and workers to 2 (or some reasonable number)? 
Similarly, what about allowing just one data file name as well (so by default training and validation is the same data set)? 

One thing I've been thinking running a little more positions for validation would give smoother curve, that's maybe without real benefit, I don't know, would ease comparison of trainings, at little extra cost.

Sopel97 commented:
so... about threads and workers. I tried on a xeon platinum a few days ago, and I don't measure any difference between 1,2,4,8. CPU utilization is not even going higher. I think 1/1 is really the best default.

> Similarly, what about allowing just one data file name as well (so by default training and validation is the same data set)?

let's leave this for a different pr

> One thing I've been thinking running a little more positions for validation would give smoother curve, that's maybe without real benefit

The dips are at least in similar places. I can read from the graphs loss differences that result in 2-3 elo at 20k nodes per move.

vondele commented:
OK, let me give it a quick testing in the current state.

vondele commented:
```
Traceback (most recent call last):
  File "train.py", line 108, in <module>
    main()
  File "train.py", line 103, in main
    train, val = data_loader_cc(args.train, args.val, feature_set, args.num_workers, batch_size, args.smart_fen_skipping, args.random_fen_skipping, main_device)
AttributeError: 'Namespace' object has no attribute 'smart_fen_skipping'
```
for
```
python train.py ../gensfen/gensfen_random_multi_pv_diff/large_gensfen_multipvdiff_100_d9.binpack  ../gensfen/gensfen_random_multi_pv_diff/large_gensfen_multipvdiff_100_d9.binpack --gpus 1 --threads 2 --num-workers 2 --batch-size 16384 --progress_bar_refresh_rate 300 --smart-fen-skipping --random-fen-skipping 3 --lambda=1.0  --max_epochs=400
```

Sopel97 commented:
right, sorry, fixed
